receiver/k8seventsreceiver/receiver.go

```diff
@@ -0,0 +1,152 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8seventsreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/k8seventsreceiver"
+
+import (
+	"context"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/obsreport"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	k8s "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+type k8seventsReceiver struct {
+	config          *Config
+	settings        component.ReceiverCreateSettings
+	client          k8s.Interface
+	logsConsumer    consumer.Logs
+	stopperChanList []chan struct{}
+	startTime       time.Time
+	ctx             context.Context
+	cancel          context.CancelFunc
+	obsrecv         *obsreport.Receiver
+}
+
+// newReceiver creates the Kubernetes events receiver with the given configuration.
+func newReceiver(
+	set component.ReceiverCreateSettings,
+	config *Config,
+	consumer consumer.Logs,
+	client k8s.Interface,
+) (component.LogsReceiver, error) {
+	transport := "http"
+
+	return &k8seventsReceiver{
+		settings:     set,
+		config:       config,
+		client:       client,
+		logsConsumer: consumer,
+		startTime:    time.Now(),
+		obsrecv: obsreport.NewReceiver(obsreport.ReceiverSettings{
+			ReceiverID:             config.ID(),
+			Transport:              transport,
+			ReceiverCreateSettings: set,
+		}),
+	}, nil
+}
+
+func (kr *k8seventsReceiver) Start(ctx context.Context, host component.Host) error {
+	kr.ctx, kr.cancel = context.WithCancel(ctx)
+
+	kr.settings.Logger.Info("starting to watch namespaces for the events.")
+	if len(kr.config.Namespaces) == 0 {
+		kr.startWatch(corev1.NamespaceAll)
+	} else {
+		for _, ns := range kr.config.Namespaces {
+			kr.startWatch(ns)
+		}
+	}
+
+	return nil
+}
+
+func (kr *k8seventsReceiver) Shutdown(context.Context) error {
+	// Stop watching all the namespaces by closing all the stopper channels.
+	for _, stopperChan := range kr.stopperChanList {
+		close(stopperChan)
+	}
+	kr.cancel()
+	return nil
+}
+
+// Add the 'New Event' handler and trigger the watch for a specific namespace.
+func (kr *k8seventsReceiver) startWatch(ns string) {
+	stopperChan := make(chan struct{})
+	kr.stopperChanList = append(kr.stopperChanList, stopperChan)
+	kr.startWatchingNamespace(kr.client, cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			ev := obj.(*corev1.Event)
+			kr.handleNewEvent(ev)
+		},
+		UpdateFunc: func(_, obj interface{}) {
+			ev := obj.(*corev1.Event)
+			kr.handleNewEvent(ev)
+		},
```

Comment:
I see we are sending one log for any update from k8s api, so I have couple questions:
1. Are we sure that any update means updated Count? If not, what other changes can be? Can you look at documentation or k8s code?
2. Can the Count be updated for example from 1 to 4? In that case we would send 1 log event which is incorrect, it should 3 logs instead. If it's too much, we may need a config option like `send_repetitive_events: true|false`.

Reply:
> 1. Are we sure that any update means updated Count? If not, what other changes can be? Can you look at the k8s documentation or code?

Yes, and a few other fields as per the [k8s source code](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/record/events_cache.go#L327).

>  2. Can the Count be updated for example from 1 to 4? In that case we would send 1 log event which is incorrect, it should be 3 logs instead. If it's too much, we may need a config option like `send_repetitive_events: true|false`.

The `Count` field is only incremented by `1`.

Reply:
Sounds good.